_posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md

```diff
@@ -0,0 +1,84 @@
+
+[Active Job](https://guides.rubyonrails.org/active_job_basics.html) is the recommended way of enqueuing and running background work in Rails. It provides a wrapper around a bunch of [different adapter implementations](https://api.rubyonrails.org/classes/ActiveJob/QueueAdapters.html) of the concept. Each implementation utilises a different underlying technology and has its own pros and cons. Most of the implementations existed before [Active Job was released](https://guides.rubyonrails.org/4_2_release_notes.html#active-job) in Rails 4.2.
+
+One of the most popular, battle-tested, and performant asynchronous frameworks is [Sidekiq](https://sidekiq.org), which is multi-threaded and utilises Redis for its queuing storage. It’s an open source library with two paid tiers for additional high-end features.
+
+
+## Instead of…
+
+…defining your jobs using Active Job:
+
+```ruby
+class DoThingsInBackgroundJob < ApplicationJob
+  queue_as :default
+
+  def peform(an_active_record_object)
+    an_active_record_object.do_things
+  end
+end
+```
+
+## Use…
+
+…Sidekiq directly:
+
+```ruby
+class DoThingsInBackgroundJob
+  include Sidekiq::Worker
+  Sidekiq_options queue: "default"
+
+  def peform(id)
+    an_active_record_object = ActiveRecordObject.find_by(id: id)
+    an_active_record_object.do_things
+  end
+end
+```
+
+
+## Why?
+
+You’re missing out on a number of advantages that you gain when you use Sidekiq directly.
+
+If you're doing _a lot_ of work in the background, enqueuing many small fast-running jobs—as is generally suggested—there is a [2—20x performance improvement](https://github.com/mperham/sidekiq/wiki/Active-Job#performance) if you use Sidekiq directly. The level of improvement will be dependant on your specific setup.
+
+If you are enqueing lots of jobs, you could be using [Sidekiq’s bulk enqueing functionality](https://github.com/mperham/sidekiq/wiki/Bulk-Queueing), this is hard to do performantly when using Sidekiq with Active Job.
+
+There are, confusingly, multiple levels of retries in the event of a job failure. Active Job has its own retry mechanism, which once completed then passes down to Sidekiq’s own (seperate and different!) retry system. This is a confusing thing to debug when things go wrong.
+
+Active Job allows passing in an Active Record object to the `#perform` method, which is then serialized to a text argument using [Global ID](https://github.com/rails/globalid). This saves doing the lookup yourself but it can raise errors if records are deleted before the job is pulled form the queue. This automatic serialization also makes the arguments to the Sidekiq jobs difficult to read in the web dashboard.
+
+There’s no real harm to “locking yourself into” a specific job framework. You are unlikely to swap to another queuing system with similarly limited Active Job functionality. When you consider replacing your asyncronous infrastructure you should assess various solutions and expect a major migration project.
```

Comment:
This paragraph could be a little clearer. The first part, I think you need to make it clearer that you're trying to say: if using Sidekiq directly, don't worry about locking yourself into something specific.

The second part: are you saying that if you're replacing your *_asynchronous_ (sp) infrastructure, then you should be considering a migration then regardless?